532/slider enhancements

eriklharper opened the pull request:
I have 3 out of the four features complete from the original issue https://github.com/Esri/calcite-components/issues/532.  These features are the top left, top right and bottom left features in this diagram:

![image](https://user-images.githubusercontent.com/821864/84432932-5f6a9300-abe2-11ea-87d9-f26500055705.png)


I need some more information about the inactive story, so I thought it best to break that out into its own ticket.

Be sure to check out the revamped demo page for this component when you run this branch: http://localhost:3333/demos/calcite-slider.html. I drastically altered it so that I could see every possible iteration and it makes for a comprehensive way of looking at the new features.

Last To-dos:
- [x] Add single value histogram shading
- [x] Fix small-value overlap issues
- [x] Prevent horizontal "jumping" effect when handles butt right up against each other before spreading further part when the hyphen appears
- [x] Prevent hyphenated range labels from extending past edge of slider in cases where the label on one side doesn't exceed the handle width, such as this case:

![image](https://user-images.githubusercontent.com/821864/84538403-71ae0500-aca6-11ea-889f-e0380b9018d6.png)

eriklharper commented on src/demos/calcite-radio-button.html:
```diff
@@ -131,135 +139,187 @@ <h3>Only one checked value (first wins, second two checked)</h3>
+                    disabled></calcite-radio-button>
+                  <calcite-radio-button theme="dark" name="firstCheckedWinsDarkAllDisabledSingle" value="3" checked>
+                  </calcite-radio-button>
 
```
☝️ changes in this file are purely formatting, I just ran prettier on it to get the HTML to look nicer.

paulcpederson commented on src/components/calcite-slider/calcite-slider.scss:
```diff
@@ -43,108 +52,195 @@ $tick-height: 4px;
+  }
+}
+.thumb--minValue {
+  transform: translate(-7px, -8px);
```
what are all these magic pixel values in the translate y position?

eriklharper replied:
The `7px` is half the handle size, so I suppose I could do `$handle-size / 2` for that and the `-8px` is the correct amount of `y` distance needed to place the center of the handle circle at the end of of the track line.  Since the handle is 7x7 you would think that the 8 could just be 7, but it needed an extra pixel for some reason.

eriklharper replied:
I found out why it has to be `8px` in the y direction: `.track { height: 2px }`.  Because the track element is set to 2px height, if you bump that down to 1 and change the transform y to 7px you'll see the circle centered on the track line, so that's why it needs to be 1 pixel more to get it to center.

paulcpederson commented on src/components/calcite-slider/calcite-slider.tsx:
```diff
@@ -102,98 +548,63 @@ export class CalciteSlider {
+          minHistogramLabeledHandle}
+
+        {!this.precise && !this.labelHandles && handle}
+        {!this.hasHistogram &&
```
can this be extracted into a `getHandle` function or something? This is really hard to read, IMO

eriklharper replied:
I agree, It is hard to read and quite frankly a mess, but I found that it would be more jumping around in the code to see what's really going on to pull them out into functions and during development of this it was handy to be able to see all the cases in one chunk of code because they all kinda inform the others in a way based on all the possible prop combinations.

paulcpederson commented on src/components/calcite-slider/calcite-slider.tsx:
```diff
@@ -211,6 +622,89 @@ export class CalciteSlider {
+    if (this.labelTicks && !this.hasHistogram && !this.isRange) {
+      return tickLabel;
+    }
+    if (
```
all of these if's return the same thing?

eriklharper replied:
Correct, this is another example of what made it easier for me in development of this portion to literally list out each specific prop combination as separate if statements.  I could probably consolidate these into a big if with `||` in between.  That would certainly cut down the visible lines of code.

eriklharper replied:
The more I think about this, I feel like combining all of these myriad combinations together into a single `if` statement will make it even messier and harder to read.  Prettier is already breaking each condition onto its own line, so to combine these all into one `if` would make probably the biggest `if` condition in javascript history, lol.

paulcpederson commented:
changes in the TSX file seem like they could be a little less repetitive/verbose. There is a _lot_ of branching in there which could get simplified.

bstifle commented:
![image](https://user-images.githubusercontent.com/14083530/84438729-e1ab8500-abeb-11ea-9e7e-7a73625c1a0a.png)
1. lets get the notch "gaps" on the sides of the notches down to 1px instead of 2px. all notches

---

![image](https://user-images.githubusercontent.com/14083530/84438988-4961d000-abec-11ea-8d93-3b8d06178d92.png)
2. If the handles are close i cant click on the handle on the right. not sure how we solve this... hit states are a bit large i guess.

---

![image](https://user-images.githubusercontent.com/14083530/84439288-cbea8f80-abec-11ea-99db-6c5a27482c15.png)
3. In general I think these hit states/focus are too large. right now they are 30x30px, should probably get closer to 18x18px square, or even 20x20



those issues said, i am LOVING these sliders. super super awesome work Erik. My def favorite component now!

eriklharper commented:
There are some issues with the Storybook display that is setting attributes like `precise="false"` which is causing some layout issues, so I'll address those in a follow-up commit.

bpatterson88 commented:
Overall really solid! Fun to interact with these.

One nit: when the step small (narrower than the width of the handle labels, I think), the rail bounds and handle labels will overlap. Example:

![Screen Shot 2020-06-11 at 2 28 10 PM](https://user-images.githubusercontent.com/15875886/84441058-f427bd80-abef-11ea-9cca-0dfbab10ffa4.png)

eriklharper commented:
Ok everyone, I have addressed all the remaining issues that have been commented on or that I also spotted myself, so if you could give this one more peek and see if this is ready to merge.  Thanks!
